near-contract-standards/src/event.rs

```diff
@@ -0,0 +1,283 @@
+    }
+
+    #[must_use = "don't forget to .emit() the event"]
+    pub fn nft_mint(data: Vec<NftMintData<'a>>) -> Self {
+        NearEvent::new_171_v1(Nep171EventKind::NftMint(data))
+    }
+
+    pub(crate) fn to_json_string(&self) -> String {
+        serde_json::to_string(self).unwrap()
+    }
+
+    pub fn to_json_event_string(&self) -> String {
+        format!("EVENT_JSON:{}", self.to_json_string())
+    }
+
+    pub fn emit(&self) {
+        near_sdk::env::log_str(&self.to_json_event_string());
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use near_sdk::AccountId;
+
+    const AUTHORIZED_ID_NONE: Option<String> = None;
+    const MEMO_NONE: Option<&String> = None;
+
+    #[test]
+    fn nft_mint() {
+        let owner_id = "bob";
+        let token_ids = vec!["0", "1"];
+        let mint_log = NftMintData::new(owner_id, token_ids, MEMO_NONE);
+        let event_log = NearEvent::nft_mint(vec![mint_log]);
+        assert_eq!(
+            serde_json::to_string(&event_log).unwrap(),
+            r#"{"standard":"nep171","version":"1.0.0","event":"nft_mint","data":[{"owner_id":"bob","token_ids":["0","1"]}]}"#
+        );
+    }
+
+    #[test]
+    fn nft_mints() {
+        let owner_id = "bob";
+        let token_ids = vec!["0", "1"];
+        let mint_log = NftMintData::new(owner_id, token_ids, MEMO_NONE);
+        let alice = AccountId::new_unchecked("alice".to_string());
+        let event_log = NearEvent::nft_mint(vec![
+            mint_log,
+            NftMintData::new(&alice, vec!["2", "3"], Some("has memo")),
+        ]);
+        assert_eq!(
+            event_log.to_json_string(),
+            r#"{"standard":"nep171","version":"1.0.0","event":"nft_mint","data":[{"owner_id":"bob","token_ids":["0","1"]},{"owner_id":"alice","token_ids":["2","3"],"memo":"has memo"}]}"#
+        );
+    }
+
+    #[test]
+    fn nft_burn() {
+        let owner_id = "bob";
+        let token_ids = vec!["0", "1"];
+        let burn_data = NftBurnData::new(owner_id, token_ids, AUTHORIZED_ID_NONE, MEMO_NONE);
+        let log = NearEvent::nft_burn(vec![burn_data]).to_json_string();
+        assert_eq!(
+            log,
+            r#"{"standard":"nep171","version":"1.0.0","event":"nft_burn","data":[{"owner_id":"bob","token_ids":["0","1"]}]}"#
+        );
+    }
+
+    #[test]
+    fn nft_burns() {
+        let owner_id = "bob";
+        let token_ids = vec!["0", "1"];
+        let log = NearEvent::nft_burn(vec![
+            NftBurnData::new("alice", vec!["2", "3"], Some("4"), Some("has memo")),
+            NftBurnData::new(owner_id, token_ids, AUTHORIZED_ID_NONE, MEMO_NONE),
+        ])
+        .to_json_string();
+        assert_eq!(
+            log,
+            r#"{"standard":"nep171","version":"1.0.0","event":"nft_burn","data":[{"owner_id":"alice","token_ids":["2","3"],"authorized_id":"4","memo":"has memo"},{"owner_id":"bob","token_ids":["0","1"]}]}"#
+        );
+    }
+
+    #[test]
+    fn nft_transfer() {
```

Comment:
Could you please also add a test with callbacks? (`nft_transfer_call`)
We need to be sure that the log was produced in the positive case, and there was no log in the negative one.

Reply:
@telezhnaya @frol @BenKurrek 

Looking into this, I see two solutions.

1) Emit the transfer after the initial transfer,  then when reverting emit the reverse transfer. (This is currently the case since `internal_transfer` is used in both cases.

2) Refrain from emitting the transfer until the end.  That is don't emit in the internal transfer, but have the caller emit the event. This seems like the more reasonable approach, however, it has a draw back with the current implementation of the standard.

Currently `nft_transfer_call` takes a `memo` parameter, to be emitted with the event.  However, if the event is emitted at the end of the chain of calls, then the final `nft_reslove_transfer` needs this and with it's current API it doesn't.

So really the choices are 1) emitted an extra log on a failure, or 2) change the standard to allow the final callback to have the `memo` argument.

Personally I think the first option is the easiest, but will end up adding 2 extra logs.  So perhaps go with the approach for now and then go through the NEP process to pass the `memo`?

Reply:
How are you going to deal with logging the `authorized_id`? See my explanation of my implementation found in [docs](https://docs.near.org/docs/tutorials/contracts/nfts/events#logging-transfers) and my code [here](https://github.com/near-examples/nft-tutorial/blob/e8f85847f93c73ac56f926315a3b6abc7290d7d1/nft-contract/src/nft_core.rs#L126)

Reply:
How are you going to deal with logging the `authorized_id`? See my implementation explained in [docs](https://docs.near.org/docs/tutorials/contracts/nfts/events#logging-transfers) and my code found [here](https://github.com/near-examples/nft-tutorial/blob/main/nft-contract/src/nft_core.rs#L126)

Reply:
Good point.  Your change isn't part of the standard either.  So adde it to my PR linked below.